### ISSUE TITLE
Default "Required" rule not using Display name attribute for client side validation.

### ISSUE THREAD

**Issue by donald-fraser** [JeremySkinner/FluentValidation#425]
FluentValidation & FluentValidation.Mvc versions: 6.2.1.0
MVC version: 5.2.3.0
.NET: 4.6.2

I have a decimal field such as:
`
        [Display(Name = "Value_key", ResourceType = typeof(BudgetPlannerRes))]
`
`        public decimal value { get; set; }
`

and the only rule I have for this field is:
`
        RuleFor( b => b.value)
                .GreaterThanOrEqualTo(0)
                .WithName(BudgetPlannerRes.Value_key);
`

The client side output is: 
`<input class="form-control" data-val="true" data-val-number="The field Value must be a number." data-val-range="&#39;Value&#39; must be greater than or equal to &#39;0&#39;." data-val-range-max="" data-val-range-min="0" data-val-required="&#39;value&#39; must not be empty." id="value" name="value" type="text" value="0.00" />`

As you can see the "data-val-required" rule is not using the resource display name. Unfortunately the NotEmpty rule, which does use the correct display resource, requires that the number not be zero so I cannot get arround this issue without resorting back to using MVC validation. The NotEmpty rule is also not consistent between client and server. The client allows a zero value whereas the server does not.

**Comment by JeremySkinner** [comment]
WithName is for simple strings, not localized resources. You need to use WithLocalizedName:

```
RuleFor( b => b.value) .GreaterThanOrEqualTo(0) .WithNameLocalizedName(() => BudgetPlannerRes.Value_key);
```

**Comment by donald-fraser** [comment]
Doesn't make any difference. The "data-val-required" rule is a rule that I am not specifying and, for what ever reason, it does not pick up the "Display" attribute. I can have no rules for that field and I still get the rule "data-val-required" being applied on the client with the incorrect field display name.

**Comment by JeremySkinner** [comment]
Sorry, I misunderstood what you were asking.

This is actually a requirement of MVC - because your property is a non-nullable value type (decimal), a required rule is automatically generated, and this can't be turned off (if you turn it off at the FV-level, then MVC just generates its own message that can't be overriden). 

There are two ways two you can handle this:
1) Either make your property nullable, and then MVC won't generate a required rule
2) Leave the property as non-nullable, but add a NotNull rule in your validator (it'll be ignored on the server-side as a decimal can never be null, but you can control the message generated on the client)

**Comment by donald-fraser** [comment]
FluentValidation is generating the message, not MVC. If I revert to MVC validation rule I get the display name as specificed by the Display attrbute.
`data-val-required="The Value field is required."`
Whereas FluentValidation the client message is:
`data-val-required="&#39;value&#39; must not be empty."`

**Comment by JeremySkinner** [comment]
Basically, MVC asks each validation provider in turn to generate the implicit message. The flow works like this:

- MVC asks FV for the message
-- FV checks if you have a NotNull/NotEmpty rule. If so, it uses that.
-- If not, it generates one (what you're seeing)
- If FV didn't generate the message, MVC asks DataAnnotations to generate one
-- DataAnnotations looks for a RequiredAttribute. If it finds one, it uses this.
-- If not, DataAnnotations generates one (with a message that can't be overriden)
- If neither FV or DataAnnotations generated the message, it generates its own (which can't be overriden)

You can turn it off at the FV level by setting AddImplicitRequiredValidator=false on your FluentValidationModelValidatorProvider where you call FluentValidationModelValidatorProvider.Configure()

```
// In your startup routine
FluentValidationModelValidatorProvider.Configure(x => x.AddImplicitREquiredValidator=false);
```

**Comment by donald-fraser** [comment]
You seem to be missing the point. I am happy that the rule is added, its just that FV is not using the Display attribute when if generates the message and it should.

**Comment by JeremySkinner** [comment]
You seem to be missing my point. I'll try and make it clearer. 

FV and DataAnnotations are separate. DataAnnotations generates its implicit rule based on the DisplayAttribute. The DisplayAttribute is part of DataAnnotations. 

The DisplayAttribute is not part of FluentValidation. FluentValidation doesn't use it. FluentValidation uses WithMessage/WithName. These are the equivalents. 

Hope that makes sense/

**Comment by JeremySkinner** [comment]
One further thing; If you wanted to add support for this in your project, I think you could do it by overriding the CreateNotNullValidatorForProperty method in the FluentValidationModelValidatorProvider. This is where the implicit rule is generated (when AddImplicitRequiredValidator is set to true), and if in here you build a PropertyRule instance that reads from the attribute I think it'd do what you want. I'll have a quick play with it and try and get you a code sample.

**Comment by JeremySkinner** [comment]
Think this should do it:

```
    public class MyCustomFVProvider : FluentValidationModelValidatorProvider {
        public MyCustomFVProvider(IValidatorFactory validatorFactory = null) : base(validatorFactory) {
        }

        protected override ModelValidator CreateNotNullValidatorForProperty(ModelMetadata metadata, ControllerContext cc) {

            var displayName = metadata.GetDisplayName();

			var fakeRule = new PropertyRule(null, x => metadata.Model, null, null, metadata.ModelType, null)
			{
				PropertyName = metadata.PropertyName,
				DisplayName = new StaticStringSource(displayName),
			};

			return new FluentValidation.Mvc.RequiredFluentValidationPropertyValidator(metadata, cc, fakeRule, new NotNullValidator());
		}
    }

```


...and as it can be done without taking a direct dependency on DataAnnotations by using GetDisplayName on the metadata instance, I'll look at getting this baked into the default provider for a future release. 

Hope that works

**Comment by donald-fraser** [comment]
Thanks for all the explanations and code example. I have used the NotNull rule and this fixes the problem. For some reason I thought that NotNull & NotEmpty were the same, but looking at the source code I reaslised they are not. 
In order for you code example to work, the method "CreateNotNullValidatorForProperty(...)" would need to be marked as virtual in the base class which it is not. 
Thanks again for you help, much appreciated.
